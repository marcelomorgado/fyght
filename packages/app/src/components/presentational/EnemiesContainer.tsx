import React, { useEffect, useState } from "react";
import { Row, Col, Divider } from "antd";
import { Enemy } from "./Enemy";
import { useFyghtContext } from "../../context";

export const EnemiesContainer: React.FC = () => {
  const [isLoading, setLoading] = useState(true);

  const {
    state: {
      enemies,
      metamask: { account },
      myFyghter,
    },
    loadEnemies,
  } = useFyghtContext();

  useEffect(() => {
    const init = async (): Promise<void> => {
      await loadEnemies();
      setLoading(false);
    };
    init();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [account, myFyghter]);

  if (isLoading) {
    return <>{`Loading ...`}</>;
  }

  if (enemies.length === 0) {
    return <>{`No enemies yet!`}</>;
  }

  return (
    <>
      <Divider orientation="left" style={{ color: "#333", fontWeight: "normal" }}>
        Enemies
      </Divider>
      <Row gutter={[16, 24]}>
        {enemies.map((enemy: Enemy, i: number) => (
          <Col key={i} className="gutter-row" span={4}>
            <Enemy enemy={enemy} />
          </Col>
        ))}
      </Row>
    </>
  );
};
