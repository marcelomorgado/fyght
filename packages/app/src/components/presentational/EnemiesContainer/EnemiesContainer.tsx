import React, { useEffect, useState } from "react";
import { Row, Col, Divider } from "antd";
import { Enemy } from "../Enemy";
import { useFyghtState } from "../../../state";

export const EnemiesContainer: React.FC = () => {
  const [isLoading, setLoading] = useState(true);

  const [
    {
      myFyghter,
      enemies,
      metamask: { ethereumAccount, networkId },
    },
    { fetchAllEnemies },
  ] = useFyghtState();

  useEffect(() => {
    const init = async (): Promise<void> => {
      await fetchAllEnemies();
      setLoading(false);
    };
    init();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [ethereumAccount, networkId, myFyghter]);

  if (isLoading) {
    return <>{`Loading ...`}</>;
  }

  if (enemies.length === 0) {
    return <>{`No enemies yet!`}</>;
  }

  return (
    <>
      <Divider orientation="left" style={{ color: "#333", fontWeight: "normal" }}>
        Enemies
      </Divider>
      <Row gutter={[16, 24]}>
        {enemies.map((enemy: Enemy, i: number) => (
          <Col key={i} className="gutter-row" span={4}>
            <Enemy enemy={enemy} myFyghter={myFyghter} />
          </Col>
        ))}
      </Row>
    </>
  );
};
