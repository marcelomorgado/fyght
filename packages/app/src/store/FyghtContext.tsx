import React, { useReducer, useContext } from "react";
import { createActions } from "./actions";
import { rootReducer, initialState } from "./reducers";

const FyghtContext = React.createContext<any | null>(null);

const asyncer = (dispatch: any, state: any) => (action: any) =>
  typeof action === "function" ? action(dispatch, state) : dispatch(action);

const FyghtProvider = ({ children }: any) => {
  const [state, baseDispatch] = useReducer(rootReducer, initialState);

  const dispatch = React.useCallback(asyncer(baseDispatch, state), []);

  const actions = createActions(dispatch, state);

  return (
    <FyghtContext.Provider value={{ state, ...actions }}>
      {children}
    </FyghtContext.Provider>
  );
};

const useFyghtContext = () => useContext(FyghtContext);

export { FyghtProvider, useFyghtContext };
