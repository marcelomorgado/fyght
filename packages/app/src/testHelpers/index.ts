import { BigNumber } from "ethers";
import { Skin } from "../constants";

export const storeMocks: { myFyghter: Fyghter; enemies: Enemy[] } = {
  myFyghter: {
    id: BigNumber.from("1"),
    name: "John",
    skin: Skin.NORMAL_GUY,
    xp: BigNumber.from("5"),
    balance: BigNumber.from("5"),
  },
  enemies: [
    {
      winProbability: BigNumber.from("50"),
      fyghter: {
        id: BigNumber.from("2"),
        name: "Charlie",
        skin: Skin.NAKED,
        xp: BigNumber.from("1"),
        balance: BigNumber.from("5"),
      },
    },
    {
      winProbability: BigNumber.from("50"),
      fyghter: {
        id: BigNumber.from("3"),
        name: "Saul",
        skin: Skin.NINJA,
        xp: BigNumber.from("10"),
        balance: BigNumber.from("5"),
      },
    },
    {
      winProbability: BigNumber.from("50"),
      fyghter: {
        id: BigNumber.from("4"),
        name: "Julian",
        skin: Skin.NAKED,
        xp: BigNumber.from("1"),
        balance: BigNumber.from("5"),
      },
    },
    {
      winProbability: BigNumber.from("50"),
      fyghter: {
        id: BigNumber.from("5"),
        name: "Brad",
        skin: Skin.MASTER,
        xp: BigNumber.from("89"),
        balance: BigNumber.from("5"),
      },
    },
    {
      winProbability: BigNumber.from("50"),
      fyghter: {
        id: BigNumber.from("6"),
        name: "Frank",
        skin: Skin.NO_ONE,
        xp: BigNumber.from("71"),
        balance: BigNumber.from("5"),
      },
    },
    {
      winProbability: BigNumber.from("50"),
      fyghter: {
        id: BigNumber.from("7"),
        name: "Brian",
        skin: Skin.MONK,
        xp: BigNumber.from("30"),
        balance: BigNumber.from("5"),
      },
    },
    {
      winProbability: BigNumber.from("50"),
      fyghter: {
        id: BigNumber.from("8"),
        name: "Fred",
        skin: Skin.KARATE_KID,
        xp: BigNumber.from("10"),
        balance: BigNumber.from("5"),
      },
    },
  ],
};
