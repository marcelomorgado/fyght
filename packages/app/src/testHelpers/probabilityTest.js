const p = (myXp, hisXp) => 1000 / (myXp + hisXp);

for (let myXp = 1; myXp <= 100; ++myXp) {
  for (let hisXp = 1; hisXp <= 100; ++hisXp) {
    console.log(`myXp=${myXp}, hisXp=${hisXp}, p=${p(myXp, hisXp)}`);
  }
}
