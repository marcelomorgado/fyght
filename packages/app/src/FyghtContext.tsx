import React, { useReducer, useContext } from "react";
import { stateMocks } from "./testHelpers";

interface FyghtContextInterface {
  myFyghter: Fyghter;
  enemies: Array<Fyghter>;
}

const FyghtContext = React.createContext<any | null>(null);

const initialState: FyghtContextInterface = {
  myFyghter: stateMocks.myFyghter,
  enemies: stateMocks.enemies,
};

const enemiesReducer = (
  state: Array<Fyghter> = initialState.enemies,
  action: { type: string; payload?: any }
): Array<Fyghter> => {
  const { type } = action;

  switch (type) {
    case "TODO":
      return state;
    default:
      return state;
  }
};

const myFyghterReducer = (
  state: Fyghter = initialState.myFyghter,
  action: { type: string; payload?: any }
): Fyghter => {
  const { type, payload } = action;
  const { name, skin } = payload;
  switch (type) {
    case "RENAME":
      return { ...state, name };
    case "CHANGE_SKIN":
      return { ...state, skin };
    default:
      return state;
  }
};

const rootReducer = (
  state: FyghtContextInterface = initialState,
  action: any
): FyghtContextInterface => {
  const { myFyghter, enemies } = state;
  return {
    myFyghter: myFyghterReducer(myFyghter, action),
    enemies: enemiesReducer(enemies, action),
  };
};

const FyghtProvider = ({ children }: any) => {
  const [state, baseDispatch] = useReducer(rootReducer, initialState);

  const dispatch = React.useCallback(asyncer(baseDispatch, state), []);

  const renameMyFyghter = (name: string) =>
    dispatch({ type: "RENAME", payload: { name } });

  const changeMyFyghterSkin = (skin: string) =>
    dispatch({ type: "CHANGE_SKIN", payload: { skin } });

  return (
    <FyghtContext.Provider
      value={{ state, renameMyFyghter, changeMyFyghterSkin }}
    >
      {children}
    </FyghtContext.Provider>
  );
};

const asyncer = (dispatch: any, state: any) => (action: any) =>
  typeof action === "function" ? action(dispatch, state) : dispatch(action);

const { Consumer: FyghtConsumer } = FyghtContext;
const useFyghtContext = () => useContext(FyghtContext);

export { FyghtProvider, FyghtConsumer, useFyghtContext };

export default FyghtContext;
