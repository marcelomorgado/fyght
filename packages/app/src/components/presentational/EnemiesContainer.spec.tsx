import * as React from "react";
import { shallow } from "enzyme";
import * as FyghterContext from "../../FyghtContext";
import { EnemiesContainer } from "./EnemiesContainer";
import { storeMocks } from "../../testHelpers";

describe("EnemiesContainer", () => {
  test("should render the component", () => {
    const contextValues = { state: { enemies: storeMocks.enemies } };

    jest
      .spyOn(FyghterContext, "useFyghtContext")
      .mockImplementation(() => contextValues);

    expect(shallow(<EnemiesContainer />)).toMatchSnapshot();
  });
});
