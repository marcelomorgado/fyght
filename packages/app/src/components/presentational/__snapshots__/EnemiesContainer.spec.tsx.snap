// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnemiesContainer after loading loading 1`] = `
<Fragment>
  <Divider
    orientation="left"
    style={
      Object {
        "color": "#333",
        "fontWeight": "normal",
      }
    }
  >
    Enemies
  </Divider>
  <Row
    gutter={
      Array [
        16,
        24,
      ]
    }
  >
    <Col
      className="gutter-row"
      key="0"
      span={4}
    >
      <Component
        enemy={
          Object {
            "id": BigNumber {
              "_hex": "0x02",
              "_isBigNumber": true,
            },
            "name": "Charlie",
            "skin": "naked",
            "xp": BigNumber {
              "_hex": "0x01",
              "_isBigNumber": true,
            },
          }
        }
      />
    </Col>
    <Col
      className="gutter-row"
      key="1"
      span={4}
    >
      <Component
        enemy={
          Object {
            "id": BigNumber {
              "_hex": "0x03",
              "_isBigNumber": true,
            },
            "name": "Saul",
            "skin": "ninja",
            "xp": BigNumber {
              "_hex": "0x0a",
              "_isBigNumber": true,
            },
          }
        }
      />
    </Col>
    <Col
      className="gutter-row"
      key="2"
      span={4}
    >
      <Component
        enemy={
          Object {
            "id": BigNumber {
              "_hex": "0x04",
              "_isBigNumber": true,
            },
            "name": "Julian",
            "skin": "naked",
            "xp": BigNumber {
              "_hex": "0x01",
              "_isBigNumber": true,
            },
          }
        }
      />
    </Col>
    <Col
      className="gutter-row"
      key="3"
      span={4}
    >
      <Component
        enemy={
          Object {
            "id": BigNumber {
              "_hex": "0x05",
              "_isBigNumber": true,
            },
            "name": "Brad",
            "skin": "master",
            "xp": BigNumber {
              "_hex": "0x59",
              "_isBigNumber": true,
            },
          }
        }
      />
    </Col>
    <Col
      className="gutter-row"
      key="4"
      span={4}
    >
      <Component
        enemy={
          Object {
            "id": BigNumber {
              "_hex": "0x06",
              "_isBigNumber": true,
            },
            "name": "Frank",
            "skin": "no_one",
            "xp": BigNumber {
              "_hex": "0x47",
              "_isBigNumber": true,
            },
          }
        }
      />
    </Col>
    <Col
      className="gutter-row"
      key="5"
      span={4}
    >
      <Component
        enemy={
          Object {
            "id": BigNumber {
              "_hex": "0x07",
              "_isBigNumber": true,
            },
            "name": "Brian",
            "skin": "monk",
            "xp": BigNumber {
              "_hex": "0x1e",
              "_isBigNumber": true,
            },
          }
        }
      />
    </Col>
    <Col
      className="gutter-row"
      key="6"
      span={4}
    >
      <Component
        enemy={
          Object {
            "id": BigNumber {
              "_hex": "0x08",
              "_isBigNumber": true,
            },
            "name": "Fred",
            "skin": "karate_kid",
            "xp": BigNumber {
              "_hex": "0x0a",
              "_isBigNumber": true,
            },
          }
        }
      />
    </Col>
  </Row>
</Fragment>
`;

exports[`EnemiesContainer loading 1`] = `
<Fragment>
  Loading ...
</Fragment>
`;
